File: typescript/src/ancilis/producers/llm.ts
/**
 * LLM SDK producers for Anthropic, OpenAI, Google Gemini, Mistral, Cohere,
 * xAI, and four OpenAI-compatible inference platforms (Groq, Together,
 * Fireworks, DeepSeek).
 *
 * Mirrors the Python `ancilis.producers.llm` module. Each producer wraps the
 * SDK call surface so every model invocation becomes an evaluated, evidence-
 * recorded Action. Duck-typed against the upstream SDKs — no hard imports.
 */

import { createHash, randomUUID } from "node:crypto";
import type { ResolvedConfig } from "../config/index.js";
import type { Action } from "../engine/action.js";
import { Engine } from "../engine/engine.js";
import { ToolRegistry, ToolStatus } from "../engine/registry.js";
import type { ToolEntry } from "../engine/registry.js";
import type { EvaluationResult } from "../engine/result.js";
import { matchesToolList } from "../engine/tool-matching.js";
import { EvidenceStore } from "../evidence/store.js";
import { canonicalJsonStringify } from "../evidence/chain.js";
import { recordAdapterUsed } from "../telemetry/index.js";
import { BlockedActionError } from "./tool.js";
import { ProducerType } from "./protocol.js";

const PRODUCER_VERSION = "0.1.0";

export interface LLMInvocation {
  model: string;
  agentName: string;
  messages?: unknown[];
  system?: unknown;
  tools?: unknown[];
  response?: unknown;
  metadata?: Record<string, unknown>;
}

export interface LLMObservation {
  action: Action;
  evaluation: EvaluationResult;
}

export interface LLMExecutionResult<R = unknown> {
  action: Action;
  evaluation: EvaluationResult;
  blocked: boolean;
  response?: R;
}

/**
 * Base producer for LLM SDK calls. Subclasses set `provider` and may override
 * `extractInvocation` to normalize provider-specific kwargs into LLMInvocation.
 */
export class LLMActionProducer {
  static readonly provider: string = "llm";

  protected _config: ResolvedConfig;
  protected _engine: Engine;
  protected _registry: ToolRegistry;
  protected _evidenceStore: EvidenceStore;
  private _sessionId: string = randomUUID();

  constructor(
    config: ResolvedConfig,
    engine: Engine,
    registry?: ToolRegistry,
    evidenceStore?: EvidenceStore,
  ) {
    this._config = config;
    this._engine = engine;
    this._registry = registry ?? engine.registry;
    this._evidenceStore = evidenceStore ?? new EvidenceStore(config);
    recordAdapterUsed(this.provider);
  }

  /** Concrete provider slug for this instance — overridden via static. */
  get provider(): string {
    return (this.constructor as typeof LLMActionProducer).provider;
  }

  get producerType(): ProducerType { return ProducerType.FRAMEWORK; }
  get producerVersion(): string { return PRODUCER_VERSION; }
  get sessionId(): string { return this._sessionId; }

  protected _toolName(invocation: LLMInvocation): string {
    const model = invocation.model || "unknown-model";
    return `llm:${this.provider}:${model}`;
  }

  protected _payload(invocation: LLMInvocation): Record<string, unknown> {
    return {
      provider: this.provider,
      model: invocation.model,
      messages: invocation.messages ?? [],
      system: invocation.system ?? null,
      tools: invocation.tools ?? [],
      metadata: invocation.metadata ?? {},
    };
  }

  protected _buildDcCodes(): string[] {
    const codes: string[] = [];
    for (const dcCodes of this._config.dataClassifications.values()) {
      for (const code of dcCodes) {
        if (!codes.includes(code)) codes.push(code);
      }
    }
    return codes;
  }

  translate(invocation: LLMInvocation): Action {
    const payload = this._payload(invocation);
    const paramHash = createHash("sha256")
      .update(canonicalJsonStringify(payload))
      .digest("hex");
    const toolName = this._toolName(invocation);
    const entry = this._registry.lookup(toolName);
    return {
      actionId: randomUUID(),
      timestamp: new Date().toISOString(),
      agentId: invocation.agentName,
      sourceType: this.producerType,
      producerType: this.producerType,
      producerVersion: this.producerVersion,
      agentOwner: this._config.agentOwner ?? null,
      actionType: "api_request",
      tool: {
        name: toolName,
        server: this.provider,
        descriptionHash: entry?.descriptionHash ?? null,
      },
      parameters: { raw: payload, parameterHash: paramHash },
      context: {
        dataClassifications: this._buildDcCodes(),
        activeOverlays: [...this._config.activeOverlays.keys()],
      },
    };
  }

  computeToolHash(toolIdentifier: string): string {
    return createHash("sha256").update(toolIdentifier).digest("hex");
  }

  registerTools(registry: ToolRegistry): string[] {
    return registry.getAll().map((entry) => entry.name);
  }

  protected _ensureRegistered(invocation: LLMInvocation): string {
    const name = this._toolName(invocation);
    if (!this._registry.lookup(name)) {
      const status = matchesToolList(name, this._config.toolsAllowed)
        ? ToolStatus.APPROVED
        : ToolStatus.OBSERVED;
      const now = new Date().toISOString();
      this._registry.register({
        name,
        descriptionHash: this.computeToolHash(name),
        status,
        approvedBy: status === ToolStatus.APPROVED ? "config" : null,
        firstSeen: now,
        statusChanged: now,
      } satisfies ToolEntry);
    }
    return name;
  }

  async observe(invocation: LLMInvocation): Promise<LLMObservation> {
    const toolName = this._ensureRegistered(invocation);
    const action = this.translate(invocation);
    const evaluation = this._engine.evaluate(action);
    await this._evidenceStore.store(evaluation, toolName);
    return { action, evaluation };
  }

  async execute<R = unknown>(
    invocation: LLMInvocation,
    transport: () => R | Promise<R>,
    enforce = false,
  ): Promise<LLMExecutionResult<R>> {
    const { action, evaluation } = await this.observe(invocation);
    if (enforce && evaluation.decision === "BLOCK") {
      throw new BlockedActionError(action.tool.name, evaluation);
    }
    const response = await transport();
    return { action, evaluation, blocked: false, response };
  }

  /**
   * Normalize provider-specific kwargs into LLMInvocation. Default extractor
   * handles the common Anthropic/OpenAI shape (`model`, `messages`, `system`,
   * `tools`); subclasses override for variant shapes.
   */
  protected extractInvocation(kwargs: Record<string, unknown>, agentName: string): LLMInvocation {
    const reserved = new Set(["model", "messages", "system", "tools"]);
    const metadata: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(kwargs)) {
      if (!reserved.has(k)) metadata[k] = v;
    }
    return {
      model: String(kwargs["model"] ?? "unknown-model"),
      agentName,
      messages: Array.isArray(kwargs["messages"]) ? (kwargs["messages"] as unknown[]) : [],
      system: kwargs["system"],
      tools: Array.isArray(kwargs["tools"]) ? (kwargs["tools"] as unknown[]) : undefined,
      metadata,
    };
  }

  /**
   * Wrap an SDK `create`-style callable so each invocation is observed first.
   * Works with `client.messages.create` (Anthropic),
   * `client.chat.completions.create` and `client.responses.create` (OpenAI),
   * and `client.models.generateContent` (Gemini, when called with kwargs).
   */
  wrapCreate<R = unknown>(
    create: (kwargs: Record<string, unknown>) => R | Promise<R>,
    agentName?: string,
    enforce = false,
  ): (kwargs: Record<string, unknown>) => Promise<LLMExecutionResult<R>> {
    return async (kwargs: Record<string, unknown>): Promise<LLMExecutionResult<R>> => {
      const invocation = this.extractInvocation(
        kwargs,
        agentName ?? this._config.agentName,
      );
      return this.execute(invocation, () => create(kwargs), enforce);
    };
  }
}

export class AnthropicActionProducer extends LLMActionProducer {
  static override readonly provider: string = "anthropic";
}

export class OpenAIActionProducer extends LLMActionProducer {
  static override readonly provider: string = "openai";

  protected override extractInvocation(
    kwargs: Record<string, unknown>,
    agentName: string,
  ): LLMInvocation {
    const reserved = new Set(["model", "messages", "system", "tools", "input", "instructions"]);
    let messages: unknown[] | undefined = Array.isArray(kwargs["messages"])
      ? (kwargs["messages"] as unknown[])
      : undefined;
    if (messages === undefined && "input" in kwargs) {
      const raw = kwargs["input"];
      if (typeof raw === "string") {
        messages = [{ role: "user", content: raw }];
      } else if (Array.isArray(raw)) {
        messages = raw;
      } else if (raw !== undefined && raw !== null) {
        messages = [raw];
      }
    }
    const metadata: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(kwargs)) {
      if (!reserved.has(k)) metadata[k] = v;
    }
    return {
      model: String(kwargs["model"] ?? "unknown-model"),
      agentName,
      messages: messages ?? [],
      system: kwargs["instructions"] ?? kwargs["system"],
      tools: Array.isArray(kwargs["tools"]) ? (kwargs["tools"] as unknown[]) : undefined,
      metadata,
    };
  }
}

export class GeminiActionProducer extends LLMActionProducer {
  static override readonly provider: string = "gemini";

  protected override extractInvocation(
    kwargs: Record<string, unknown>,
    agentName: string,
  ): LLMInvocation {
    const reserved = new Set(["model", "contents", "systemInstruction", "system_instruction", "tools", "config"]);
    const contents = kwargs["contents"];
    let messages: unknown[] = [];
    if (typeof contents === "string") {
      messages = [{ role: "user", content: contents }];
    } else if (Array.isArray(contents)) {
      messages = contents;
    } else if (contents !== undefined && contents !== null) {
      messages = [contents];
    }

    let system: unknown;
    let tools: unknown[] | undefined;
    const config = kwargs["config"];
    if (config && typeof config === "object" && !Array.isArray(config)) {
      const cfg = config as Record<string, unknown>;
      system = cfg["system_instruction"] ?? cfg["systemInstruction"] ?? kwargs["system_instruction"];
      tools = Array.isArray(cfg["tools"])
        ? (cfg["tools"] as unknown[])
        : Array.isArray(kwargs["tools"])
          ? (kwargs["tools"] as unknown[])
          : undefined;
    } else {
      system = kwargs["system_instruction"] ?? kwargs["systemInstruction"];
      tools = Array.isArray(kwargs["tools"]) ? (kwargs["tools"] as unknown[]) : undefined;
    }

    const metadata: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(kwargs)) {
      if (!reserved.has(k)) metadata[k] = v;
    }
    return {
      model: String(kwargs["model"] ?? "unknown-model"),
      agentName,
      messages,
      system,
      tools,
      metadata,
    };
  }
}

export class MistralActionProducer extends LLMActionProducer {
  static override readonly provider: string = "mistral";
}

export class CohereActionProducer extends LLMActionProducer {
  static override readonly provider: string = "cohere";

  protected override extractInvocation(
    kwargs: Record<string, unknown>,
    agentName: string,
  ): LLMInvocation {
    const reserved = new Set([
      "model",
      "messages",
      "message",
      "chat_history",
      "chatHistory",
      "preamble",
      "system",
      "tools",
    ]);
    let messages: unknown[] = Array.isArray(kwargs["messages"])
      ? [...(kwargs["messages"] as unknown[])]
      : [];
    if (messages.length === 0) {
      const history = kwargs["chat_history"] ?? kwargs["chatHistory"];
      if (Array.isArray(history)) messages.push(...(history as unknown[]));
      const current = kwargs["message"];
      if (current !== undefined && current !== null) {
        messages.push({ role: "user", content: current });
      }
    }
    const metadata: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(kwargs)) {
      if (!reserved.has(k)) metadata[k] = v;
    }
    return {
      model: String(kwargs["model"] ?? "unknown-model"),
      agentName,
      messages,
      system: kwargs["preamble"] ?? kwargs["system"],
      tools: Array.isArray(kwargs["tools"]) ? (kwargs["tools"] as unknown[]) : undefined,
      metadata,
    };
  }
}

// xAI / Groq / Together / Fireworks / DeepSeek all expose OpenAI-compatible
// chat APIs. They subclass OpenAIActionProducer to inherit the input/messages
// normalization and change only the provider slug.

export class XAIActionProducer extends OpenAIActionProducer {
  static override readonly provider: string = "xai";
}

export class GroqActionProducer extends OpenAIActionProducer {
  static override readonly provider: string = "groq";
}

export class TogetherActionProducer extends OpenAIActionProducer {
  static override readonly provider: string = "together";
}

export class FireworksActionProducer extends OpenAIActionProducer {
  static override readonly provider: string = "fireworks";
}

export class DeepSeekActionProducer extends OpenAIActionProducer {
  static override readonly provider: string = "deepseek";
}

File: python/src/ancilis/producers/llm.py
"""LLM SDK producers for Anthropic, OpenAI, and Google Gemini.

Producers wrap LLM SDK calls in the user's runtime so each invocation becomes
an Action object the engine can evaluate. Mirrors HTTPActionProducer in shape:
observe-first, optional enforce, optional client wrapping. Duck-typed against
the upstream SDKs — the producer never imports anthropic / openai / google-genai,
so it works whether or not the SDK is installed.
"""

from __future__ import annotations

import functools
import hashlib
import json
import uuid
from collections.abc import Callable, Iterable, Mapping
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any, ParamSpec, TypeVar

from ancilis.config import ResolvedConfig
from ancilis.engine.action import Action, ActionContext, ActionParameters, ToolInfo
from ancilis.engine.engine import Engine
from ancilis.engine.registry import ToolEntry, ToolRegistry, ToolStatus
from ancilis.engine.result import EvaluationResult
from ancilis.evidence.store import EvidenceStore
from ancilis.producers.protocol import ProducerType
from ancilis.producers.tool import BlockedActionError
from ancilis.telemetry import record_adapter_used

P = ParamSpec("P")
R = TypeVar("R")


@dataclass
class LLMInvocation:
    """Provider-agnostic representation of an LLM SDK call."""

    model: str
    agent_name: str
    messages: list[Any] = field(default_factory=list)
    system: Any = None
    tools: list[Any] | None = None
    response: Any = None
    metadata: dict[str, Any] = field(default_factory=dict)


@dataclass
class LLMObservation:
    action: Action
    evaluation: EvaluationResult


@dataclass
class LLMExecutionResult:
    action: Action
    evaluation: EvaluationResult
    blocked: bool
    response: Any = None


class LLMActionProducer:
    """Base producer for LLM SDK calls.

    Subclasses set ``provider`` and may override ``_extract_invocation`` to
    normalize provider-specific kwargs into ``LLMInvocation``. The default
    extractor handles the common Anthropic/OpenAI shape (``model``,
    ``messages``, ``system``, ``tools``).
    """

    provider: str = "llm"

    def __init__(
        self,
        config: ResolvedConfig,
        engine: Engine,
        registry: ToolRegistry | None = None,
        evidence_store: EvidenceStore | None = None,
    ) -> None:
        self._config = config
        self._engine = engine
        self._registry = registry or engine.registry
        self._evidence_store = evidence_store if evidence_store is not None else EvidenceStore(config)
        self._session_id = str(uuid.uuid4())
        record_adapter_used(self.provider)

    @property
    def session_id(self) -> str:
        return self._session_id

    @property
    def producer_type(self) -> ProducerType:
        return ProducerType.FRAMEWORK

    @property
    def producer_version(self) -> str:
        return "0.1.0"

    def _tool_name(self, invocation: LLMInvocation) -> str:
        model = invocation.model or "unknown-model"
        return f"llm:{self.provider}:{model}"

    def _payload(self, invocation: LLMInvocation) -> dict[str, Any]:
        return {
            "provider": self.provider,
            "model": invocation.model,
            "messages": invocation.messages,
            "system": invocation.system,
            "tools": invocation.tools or [],
            "metadata": invocation.metadata,
        }

    def translate(self, raw_invocation: LLMInvocation) -> Action:
        payload = self._payload(raw_invocation)
        param_hash = hashlib.sha256(
            json.dumps(payload, sort_keys=True, default=repr).encode()
        ).hexdigest()
        tool_name = self._tool_name(raw_invocation)
        entry = self._registry.lookup(tool_name)
        dc_codes: list[str] = []
        for codes in self._config.data_classifications.values():
            for code in codes:
                if code not in dc_codes:
                    dc_codes.append(code)
        return Action(
            action_id=str(uuid.uuid4()),
            timestamp=datetime.now(timezone.utc).isoformat(),
            agent_id=raw_invocation.agent_name,
            source_type=self.producer_type.value,
            agent_owner=self._config.agent_owner or None,
            action_type="api_request",
            tool=ToolInfo(
                name=tool_name,
                server=self.provider,
                description_hash=entry.description_hash if entry else None,
            ),
            parameters=ActionParameters(raw=payload, parameter_hash=param_hash),
            context=ActionContext(
                data_classifications=dc_codes,
                active_overlays=list(self._config.active_overlays.keys()),
            ),
            producer_type=self.producer_type.value,
            producer_version=self.producer_version,
        )

    def compute_tool_hash(self, tool_identifier: Any) -> str:
        return hashlib.sha256(str(tool_identifier).encode()).hexdigest()

    def register_tools(self, registry: ToolRegistry) -> list[str]:
        return [entry.name for entry in registry.get_all()]

    def _ensure_registered(self, invocation: LLMInvocation) -> str:
        tool_name = self._tool_name(invocation)
        if self._registry.lookup(tool_name) is None:
            status = (
                ToolStatus.APPROVED
                if tool_name in self._config.tools_allowed
                else ToolStatus.OBSERVED
            )
            self._registry.register(
                ToolEntry(
                    name=tool_name,
                    description_hash=self.compute_tool_hash(tool_name),
                    status=status,
                    approved_by="config" if status == ToolStatus.APPROVED else None,
                )
            )
        return tool_name

    def observe(self, invocation: LLMInvocation) -> LLMObservation:
        tool_name = self._ensure_registered(invocation)
        action = self.translate(invocation)
        evaluation = self._engine.evaluate(action)
        self._evidence_store.store(evaluation, tool_name=tool_name)
        return LLMObservation(action=action, evaluation=evaluation)

    def execute(
        self,
        invocation: LLMInvocation,
        *,
        transport: Callable[P, R],
        transport_args: tuple[Any, ...] = (),
        transport_kwargs: dict[str, Any] | None = None,
        enforce: bool = False,
    ) -> LLMExecutionResult:
        observation = self.observe(invocation)
        if enforce and observation.evaluation.decision == "BLOCK":
            raise BlockedActionError(observation.action.tool.name, observation.evaluation)
        response = transport(*transport_args, **(transport_kwargs or {}))
        return LLMExecutionResult(
            action=observation.action,
            evaluation=observation.evaluation,
            blocked=False,
            response=response,
        )

    def _extract_invocation(
        self, kwargs: Mapping[str, Any], *, agent_name: str
    ) -> LLMInvocation:
        return LLMInvocation(
            model=str(kwargs.get("model") or "unknown-model"),
            agent_name=agent_name,
            messages=list(kwargs.get("messages") or []),
            system=kwargs.get("system"),
            tools=list(kwargs.get("tools") or []) or None,
            metadata={
                k: v
                for k, v in kwargs.items()
                if k not in {"model", "messages", "system", "tools"}
            },
        )

    def wrap_create(
        self,
        create: Callable[..., Any],
        *,
        agent_name: str | None = None,
        enforce: bool = False,
    ) -> Callable[..., LLMExecutionResult]:
        """Wrap an SDK ``create``-style callable so each invocation is observed.

        Works with ``client.messages.create`` (Anthropic),
        ``client.chat.completions.create`` and ``client.responses.create``
        (OpenAI), and ``client.models.generate_content`` (Gemini, when called
        with keyword args).
        """

        @functools.wraps(create)
        def wrapped(*args: Any, **kwargs: Any) -> LLMExecutionResult:
            invocation = self._extract_invocation(
                kwargs, agent_name=agent_name or self._config.agent_name
            )
            return self.execute(
                invocation,
                transport=create,
                transport_args=args,
                transport_kwargs=kwargs,
                enforce=enforce,
            )

        return wrapped


class AnthropicActionProducer(LLMActionProducer):
    """Producer for the Anthropic Python SDK (``messages.create``)."""

    provider = "anthropic"


class OpenAIActionProducer(LLMActionProducer):
    """Producer for the OpenAI Python SDK.

    Handles both ``chat.completions.create`` (``messages``) and
    ``responses.create`` (``input``) by normalizing ``input`` to ``messages``
    when present.
    """

    provider = "openai"

    def _extract_invocation(
        self, kwargs: Mapping[str, Any], *, agent_name: str
    ) -> LLMInvocation:
        messages = kwargs.get("messages")
        if messages is None and "input" in kwargs:
            raw_input = kwargs["input"]
            if isinstance(raw_input, str):
                messages = [{"role": "user", "content": raw_input}]
            elif isinstance(raw_input, Iterable):
                messages = list(raw_input)
            else:
                messages = [raw_input]
        return LLMInvocation(
            model=str(kwargs.get("model") or "unknown-model"),
            agent_name=agent_name,
            messages=list(messages or []),
            system=kwargs.get("instructions") or kwargs.get("system"),
            tools=list(kwargs.get("tools") or []) or None,
            metadata={
                k: v
                for k, v in kwargs.items()
                if k not in {"model", "messages", "system", "tools", "input", "instructions"}
            },
        )


class GeminiActionProducer(LLMActionProducer):
    """Producer for the Google Gemini SDK (``google-genai``).

    Normalizes ``contents`` (Gemini) to ``messages`` and exposes
    ``system_instruction`` as ``system``.
    """

    provider = "gemini"

    def _extract_invocation(
        self, kwargs: Mapping[str, Any], *, agent_name: str
    ) -> LLMInvocation:
        contents = kwargs.get("contents")
        if isinstance(contents, str):
            messages: list[Any] = [{"role": "user", "content": contents}]
        elif isinstance(contents, Iterable):
            messages = list(contents)
        else:
            messages = [contents] if contents is not None else []
        config = kwargs.get("config") or {}
        if isinstance(config, Mapping):
            system = config.get("system_instruction") or kwargs.get("system_instruction")
            tools = config.get("tools") or kwargs.get("tools")
        else:
            system = kwargs.get("system_instruction")
            tools = kwargs.get("tools")
        return LLMInvocation(
            model=str(kwargs.get("model") or "unknown-model"),
            agent_name=agent_name,
            messages=messages,
            system=system,
            tools=list(tools or []) or None,
            metadata={
                k: v
                for k, v in kwargs.items()
                if k not in {"model", "contents", "system_instruction", "tools", "config"}
            },
        )


class MistralActionProducer(LLMActionProducer):
    """Producer for the Mistral La Plateforme SDK (``mistralai``).

    Mistral's ``client.chat.complete(model, messages, ...)`` shape mirrors
    OpenAI/Anthropic, so the base extractor covers it.
    """

    provider = "mistral"


class CohereActionProducer(LLMActionProducer):
    """Producer for the Cohere SDK.

    Cohere's chat API uses ``message`` (single string) and ``chat_history``
    (list of role/text dicts). This extractor folds both into ``messages``
    so downstream evaluators see the same shape as other providers.
    """

    provider = "cohere"

    def _extract_invocation(
        self, kwargs: Mapping[str, Any], *, agent_name: str
    ) -> LLMInvocation:
        messages = list(kwargs.get("messages") or [])
        if not messages:
            history = kwargs.get("chat_history") or []
            if isinstance(history, Iterable):
                messages = list(history)
            current = kwargs.get("message")
            if current is not None:
                messages.append({"role": "user", "content": current})
        return LLMInvocation(
            model=str(kwargs.get("model") or "unknown-model"),
            agent_name=agent_name,
            messages=messages,
            system=kwargs.get("preamble") or kwargs.get("system"),
            tools=list(kwargs.get("tools") or []) or None,
            metadata={
                k: v
                for k, v in kwargs.items()
                if k
                not in {
                    "model",
                    "messages",
                    "message",
                    "chat_history",
                    "preamble",
                    "system",
                    "tools",
                }
            },
        )


class XAIActionProducer(OpenAIActionProducer):
    """Producer for the xAI Grok API.

    xAI exposes an OpenAI-compatible chat API, so this subclasses
    ``OpenAIActionProducer`` and changes only the provider slug.
    """

    provider = "xai"


# --- OpenAI-compatible serverless inference platforms ---
#
# Per Q2 2026 research, the consolidated serverless inference market
# (Together, Fireworks, Anyscale, Groq, Cerebras, Replicate, OctoAI) all
# expose OpenAI-compatible endpoints. The producers below are thin
# subclasses that change only the provider slug so evidence is correctly
# attributed; they reuse the OpenAI extractor for messages/input handling.


class GroqActionProducer(OpenAIActionProducer):
    """Producer for Groq's LPU-backed OpenAI-compatible API."""

    provider = "groq"


class TogetherActionProducer(OpenAIActionProducer):
    """Producer for Together AI's OpenAI-compatible inference API."""

    provider = "together"


class FireworksActionProducer(OpenAIActionProducer):
    """Producer for Fireworks AI's OpenAI-compatible inference API."""

    provider = "fireworks"


class DeepSeekActionProducer(OpenAIActionProducer):
    """Producer for the DeepSeek OpenAI-compatible API."""

    provider = "deepseek"
